include/xtensor-sparse/xcsf_scheme.hpp

```diff
@@ -0,0 +1,257 @@
+        using storage_type = ST;
+        using index_type = IT;
+
+        using value_type = typename storage_type::value_type;
+        using const_reference = typename storage_type::const_reference;
+        using pointer = typename storage_type::pointer;
+
+        const position_type& position() const;
+        const coordinate_type& coordinate() const;
+        const storage_type& storage() const;
+
+        pointer find_element(const index_type& index);
+        void insert_element(const index_type& index, const_reference value);
+        void remove_element(const index_type& index);
+
+        template<class strides_type>
+        void update_entries(const strides_type& old_strides,
+                            const strides_type& new_strides);
+    private:
+
+        position_type m_pos;
+        coordinate_type m_coords;
+        storage_type m_storage;
+    };
+
+    /******************************
+     * xcsf_scheme implementation *
+     ******************************/
+
+    template <class P, class C, class ST, class IT>
+    inline auto xcsf_scheme<P, C, ST, IT>::position() const -> const position_type&
+    {
+        return m_pos;
+    }
+
+    template <class P, class C, class ST, class IT>
+    inline auto xcsf_scheme<P, C, ST, IT>::coordinate() const -> const coordinate_type&
+    {
+        return m_coords;
+    }
+
+    template <class P, class C, class ST, class IT>
+    inline auto xcsf_scheme<P, C, ST, IT>::storage() const -> const storage_type&
+    {
+        return m_storage;
+    }
+
+    template <class P, class C, class ST, class IT>
+    inline auto xcsf_scheme<P, C, ST, IT>::find_element(const index_type& index) -> pointer
+    {
+        if (m_pos.size() == 0)
+        {
+            return nullptr;
+        }
+
+        std::size_t ielem = 0;
+        for(std::size_t i=0; i<index.size(); ++i)
+        {
+            auto it = std::find(m_coords[i].cbegin() + m_pos[i][ielem], m_coords[i].cbegin() + m_pos[i][ielem + 1], index[i]);
+            if (it != m_coords[i].cbegin() + m_pos[i][ielem + 1])
+            {
+                if (i == index.size() - 1)
+                {
+                    std::ptrdiff_t dst = std::distance(m_coords[i].cbegin(), it);
+                    return &(*(m_storage.begin() + dst));
+                }
+                else
+                {
+                    ielem = it - (m_coords[i].cbegin() + m_pos[i][ielem]);
+                }
+            }
+            else
+            {
+                return nullptr;
+            }
+        }
+        return &m_storage[ielem];
+    }
+
+    template <class P, class C, class ST, class IT>
+    inline void xcsf_scheme<P, C, ST, IT>::insert_element(const index_type& index, const_reference value)
+    {
+        auto ielem = detail::insert_index(m_pos, m_coords, index);
+        if (ielem == std::numeric_limits<std::size_t>::max())
+        {
+            throw std::runtime_error("This should not happen");
+        }
```

Comment:
Maybe we should replace this with an assert to avoid performance hit. Or reuse xtensor assertions.